### PR TITLE
Convert tuples to structs in Stake rewards-calculation functions

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
It's difficult to remember what the return values from the Stake program's internal rewards-calculation functions signify. This one is especially awful:
```
fn calculate_stake_points_and_credits( ...) -> Option<(u64, u64, u64)> 
```

#### Summary of Changes
Replace opaque tuples with helper structs
No behavior changes
Open to bikeshedding on struct field names.

### CODE REVIEW

File: programs/stake/src/stake_state.rs [hunk 1]
```diff
@@ -209,7 +215,14 @@ fn calculate_stake_points(
         true, // this is safe because this flag shouldn't affect the first
               // element (i.e. `points`) of the returned tuple in any way
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:216]: nit: outdated comment after this pr's nice cleanup. :)

File: programs/stake/src/stake_state.rs [hunk 2]
```diff
@@ -182,17 +182,23 @@ fn redeem_stake_rewards(
...
-        stake.credits_observed = credits_observed;
-        stake.delegation.stake += stakers_reward;
-        (stakers_reward, voters_reward)
-    })
+    .map(
+        |StakeRewardDistributions {
+             staker_rewards,
+             voter_rewards,
+             credits_observed,
+         }| {
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:190]: nits: this might be personal taste, but i'd like to see a more nuanced rewrite than this verbatim version like below:

```rust
.map(|stake_reward_distributions| {
  ...
  stale.credits_observed = stake_reward_distributions.credits_observed;
  ...
})
```

(or destructure `stake_reward_distributions`if this looks too wordy, java-esque... xD)

imo, large destructuring in params causes some reading pauses and makes it easy to forget the context of unqualified identifiers like `credits_observed` after that.

File: programs/stake/src/stake_state.rs [hunk 3]
```diff
@@ -209,7 +215,14 @@ fn calculate_stake_points(
         true, // this is safe because this flag shouldn't affect the first
               // element (i.e. `points`) of the returned tuple in any way
     )
-    .0
+    .points
+}
+
+#[derive(Debug, PartialEq)]
+struct StakePointsEarned {
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:222]: bikeshed: `EarnedStakePoints` or `RewardedStakePoints` or `CalculatedStakePoints`?  (it seems that `earned` and `rewarded` are already co-existing...)

I want to strive for normal english grammartical order for overall consystency across this mono-repo. speaking of which, i think `credits_observed` is a outlier and should be `observed_credits` actually....

File: programs/stake/src/stake_state.rs [hunk 4]
```diff
@@ -303,7 +324,18 @@ fn calculate_stake_points_and_credits(
...
-    (points, new_credits_observed, false)
+    StakePointsEarned {
+        points,
+        new_credits_observed,
+        force_credits_update_with_skipped_reward: false,
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct StakeRewardDistributions {
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:335]: bikeshed: how about `DistributedStakeRewards` or `CalculatedStakeRewards`?

File: programs/stake/src/stake_state.rs [hunk 5]
```diff
@@ -303,7 +324,18 @@ fn calculate_stake_points_and_credits(
...
+    staker_rewards: u64,
+    voter_rewards: u64,
+    credits_observed: u64,
```
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:338]: nits: maybe better off inheriting the `new_` prefix from `StakePointsEarned` for consistency and its actual usage?

**Review by ryoqun — APPROVED** [review]
lgtm with nits. really appreciate spending time for this cleanup for benefits of all future readers.

**Comment by CriesofCarrots** [comment]
Thanks for all the good ideas, @ryoqun !

File: programs/stake/src/stake_state.rs (hunk 3 above)
Comment (CriesofCarrots) [reply programs/stake/src/stake_state.rs:222]: I agree, but I think we're kind of locked into `credits_observed` in the `Stake` struct. I played with different options for the credits field in these structs, but ultimately decided to keep the word order consistent with Stake: `new_credits_observed`

File: programs/stake/src/stake_state.rs (hunk 4 above)
Comment (CriesofCarrots) [reply programs/stake/src/stake_state.rs]: I went with the `Calculated` prefix for both structs

**Review by t-nelson — APPROVED** [review]
aw. when did we draw straws? :slightly_smiling_face: 

lgtm!  thanks for being the one to finally bite the bullet!
